lib/proxy/peer/quicserver.go

```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+	// runCancel cancels runCtx.
+	runCancel context.CancelFunc
+	// serveCtx is a context that gets canceled when all listeners should stop
+	// accepting new connections.
+	serveCtx context.Context
+	// serveCancel cancels serveCtx.
+	serveCancel context.CancelFunc
+}
 
 // NewQUICServer returns a [QUICServer] with the given config.
 func NewQUICServer(cfg QUICServerConfig) (*QUICServer, error) {
 	if err := cfg.checkAndSetDefaults(); err != nil {
 		return nil, trace.Wrap(err)
 	}
-	panic("QUIC proxy peering is not implemented")
+
+	tlsConfig := utils.TLSConfig(cfg.CipherSuites)
+	tlsConfig.GetCertificate = cfg.GetCertificate
+	tlsConfig.VerifyPeerCertificate = verifyPeerCertificateIsProxy
+	tlsConfig.NextProtos = []string{quicNextProto}
+	tlsConfig.ClientAuth = tls.RequireAndVerifyClientCert
+	tlsConfig.MinVersion = tls.VersionTLS13
+
+	getClientCAs := cfg.GetClientCAs
+	tlsConfig.GetConfigForClient = func(chi *tls.ClientHelloInfo) (*tls.Config, error) {
+		clientCAs, err := getClientCAs(chi)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		utils.RefreshTLSConfigTickets(tlsConfig)
+		c := tlsConfig.Clone()
+		c.ClientCAs = clientCAs
+		return c, nil
+	}
+
+	quicConfig := &quic.Config{
+		MaxStreamReceiveWindow:     quicMaxReceiveWindow,
+		MaxConnectionReceiveWindow: quicMaxReceiveWindow,
+
+		MaxIncomingStreams:    1 << 60,
+		MaxIncomingUniStreams: -1,
+
+		MaxIdleTimeout:  quicMaxIdleTimeout,
+		KeepAlivePeriod: quicKeepAlivePeriod,
+
+		Allow0RTT: true,
+	}
+
+	runCtx, runCancel := context.WithCancel(context.Background())
+	serveCtx, serveCancel := context.WithCancel(runCtx)
+
+	return &QUICServer{
+		log:           cfg.Log,
+		clusterDialer: cfg.ClusterDialer,
+		tlsConfig:     tlsConfig,
+		quicConfig:    quicConfig,
+
+		runCtx:      runCtx,
+		runCancel:   runCancel,
+		serveCtx:    serveCtx,
+		serveCancel: serveCancel,
+	}, nil
 }
 
 // Serve opens a listener and serves incoming connection. Returns after calling
 // Close or Shutdown.
 func (s *QUICServer) Serve(t *quic.Transport) error {
-	panic("QUIC proxy peering is not implemented")
+	s.mu.Lock()
+	if s.closed {
+		s.mu.Unlock()
+		return trace.Wrap(quic.ErrServerClosed)
+	}
+	s.wg.Add(1)
+	defer s.wg.Done()
+	s.mu.Unlock()
+
+	lis, err := t.ListenEarly(s.tlsConfig, s.quicConfig)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer lis.Close()
+	defer context.AfterFunc(s.serveCtx, func() { _ = lis.Close() })()
+
+	for {
+		// the listener will be closed when serveCtx is done, but Accept will
+		// return any queued connection before erroring out with a
+		// [quic.ErrServerClosed]
+		c, err := lis.Accept(context.Background())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		s.wg.Add(1)
+		go s.handleConn(c)
+	}
+}
+
+func (s *QUICServer) handleConn(c quic.EarlyConnection) {
+	defer s.wg.Done()
+
+	log := s.log.With(
+		"remote_addr", c.RemoteAddr().String(),
+		"internal_id", uuid.NewString(),
+	)
+	state := c.ConnectionState()
+	log.InfoContext(s.serveCtx,
```

Comment:
Start is logged is logged using s.serveCtx, connection closed is logged using c.Context() (line 220), down below we do AcceptStream using context.Background(). I'm not sure I follow the logic, shouldn't all of them either use s.serveCtx or c.Context() ?

Reply:
(see answer to https://github.com/gravitational/teleport/pull/47587#discussion_r1803818739)

Reply:
(`AcceptStream` should run until the connection is closed, anyway - we don't currently have a way to gracefully reject streams, I'd like for quic-go to get the ability to alter the MaxIncomingStream value on a running connection first, so we can set that to 0 and then consume all the queued streams)